[P132] Additional support for INA226, INA228, INA230, INA231 and INA260

tonhuisman opened the pull request:
Resolves #2622
Resolves #2602
Forum request: https://www.letscontrolit.com/forum/viewtopic.php?t=10746
Forum request: https://www.letscontrolit.com/forum/viewtopic.php?p=74843

Features:
- [P132] Add support for:
  - INA219 (Is already supported in P027, with more configuration options, it's also included in this library)
  - INA226
  - INA228 ~(waiting for hardware to arrive)~
  - INA230 (no hardware available)
  - INA231 (often detected as INA226, but has lower specs)
  - INA260 (built-in shunt resistor)
- By default only enabled for ESP32, but can be enabled in an ESP8266 Custom build

TODO:
- [x] Update documentation
- [x] Testing!
  - [x] INA3221
  - [x] INA219
  - [x] INA226
  - [x] INA228
  - INA230 (not able to test for no hardware available)
  - [x] INA231 (is detected as INA226)
  - [x] INA260 ~(waiting for hardware to arrive)~
- [x] Refactor 'Max. current' and 'Shunt resistor' settings for higher resolution and better match.

tonhuisman committed:
[P132] Additional support for INA226, INA228, INA230, INA231 and INA260

tonhuisman committed:
[P132] Fix some typos

tonhuisman committed:
[P132] Fix some more typos

tonhuisman committed:
[P132] Fix one more typo and clean up code

tonhuisman committed:
[P132] Add INA231 to plugin name

tonhuisman committed:
[P132] Update documentation

Gonzol999 commented:
Great job, thank you very much, I'm interested in INA228 support myself. Will there be binaries available for download for the ESP32?

tonhuisman commented:
> Will there be binaries available for download for the ESP32?

Assuming you mean ESP32-Classic, here are the builds that this plugin is included in:

[ESP_Easy_mega_20251228_energy_ESP32_4M316k-2.zip](https://github.com/user-attachments/files/24360002/ESP_Easy_mega_20251228_energy_ESP32_4M316k-2.zip) for ESP32 with 4MB Flash
[ESP_Easy_mega_20251228_max_ESP32_16M8M-2.zip](https://github.com/user-attachments/files/24360004/ESP_Easy_mega_20251228_max_ESP32_16M8M-2.zip) for ESP32 with 16MB Flash

Both include the OTA `.bin` and the Serial-flashable `.factory.bin` (address 0x0).

Edit: Re-uploaded the zip files (appended `-2`) with the adjusted Conversion rate settings I pushed a few minutes ago.

NB: Possibly the INA228 Power setting might not return the expected result, as the code shows a TODO there... 🤔

tonhuisman committed:
[P132] Fix conversion rate settings for INA228

Gonzol999 commented:
Thanks for the binaries. I loaded the 4M316k.factory module, and it boots up, but I doubt it will work correctly. There's no option to select the shunt value. Without checking the documentation, I'm not entirely sure if all the INA3221/INA226/INA228/INA231/INA260 models have the same registers. At first glance, the INA3221 is a 3-channel model, while the others aren't. So how would this work without selecting the INA model in the plugin?

tonhuisman commented:
> So how would this work without selecting the INA model in the plugin?

Not sure what/how you installed the binary, but the P132 plugin ("Energy (DC) - INA3221/INA226/INA228/INA231/INA260") should show an `INA type` selector, where you can choose what INA chip is used on the board, and ditto for the `Shunt resistor` value (that selector was already available in the original version of the plugin, and now has 2 extra values added).
Can you show a screenshot of the device configuration for the plugin? Does it show the name of the build near the bottom-right end of the web page? It should be of today, 20251228, as I made a fresh build (also reflected in the name of the .zip files).

Gonzol999 commented:
<img width="1366" height="768" alt="Config ESP32" src="https://github.com/user-attachments/assets/b28702df-99ee-4f3e-a7af-00e861a63c82" />
<img width="1366" height="768" alt="Devices INA228_1" src="https://github.com/user-attachments/assets/055e3882-7c91-4934-86c1-69432303be8d" />
<img width="1366" height="768" alt="Devices INA228_2" src="https://github.com/user-attachments/assets/0f9b2cad-a53b-4e14-96f5-0c96b108ba45" />

tonhuisman commented:
Ah, this is a phenomenon we sometimes see when some other firmware (ESPHome, Espruna, Tasmota?) has been used on the esp, and the flash hasn't been wiped before installing ESPEasy. It tries to re-use an existing file system, but if that's too small it fails in writing the configuration file(s), causing weird behavior and/or crashed like you have shown.
Solution: First wipe the entire flash, you can do that using the web-flasher at https://td-er.nl/ESPEasy/latest, flash the last Energy build on it, connect to WiFi as usual and then install the above build via OTA (Tools/Firmware update) using the .bin file (_not_ the .factory.bin file). Your current configuration is crooked anyway, so starting from scratch is the best way to solve that.

tonhuisman committed:
[P132] Fix defaults for INA228

tonhuisman committed:
[P132] Update documentation for INA228 conversion rates

Gonzol999 commented:
Thank you again for your involvement and willingness to help. It's possible this is a hardware issue with my ESP32; none of the methods I know of for clearing the memory have yielded positive results. Since I only have one ESP32 board (I have many, but they're ESP8266s), I've ordered another ESP32, but it will probably take some time for them to arrive.

tonhuisman commented:
> I have many, but they're ESP8266s

I've built you a Custom Energy build with this feature enabled for ESP8266 with 4MB Flash:
[ESP_Easy_mega_20251229_energy_ESP8266_4M1M.zip](https://github.com/user-attachments/files/24371014/ESP_Easy_mega_20251229_energy_ESP8266_4M1M.zip)
(The .bin.gz is a gzipped version of the .bin file, for faster upload via OTA, allow about 20 seconds after the OTA upload has finished so the data is unpacked in the correct flash location)

Gonzol999 commented:
Another huge thank you. The ESP8266 binary is better. Almost everything seems to work, but there's a problem with reading data. Admittedly, I've been experimenting with the INA226 for a while now. I have a question: instead of selecting the internal resistor value, there should be an option to select a shunt, e.g., 30A (75mV), etc., because these built-in resistors are intended for small pads.

tonhuisman commented:
Selecting a different shunt resistor is mostly a wiring issue, finding the right shunt value (for 30A the 2 mOhm setting should be just right for an INA226), find a 0.002 Ohm resistor that can handle the power dissipation, remove the existing shunt from the board and wire the new shunt to the board.
Currently 0.002 Ohm is the smallest shunt supported, (and previously 0.005 Ohm was the smallest, allowing ca. 16A), but nobody ever asked for smaller or more flexible configuration (the plugin was available in the [ESPEasy PluginPlayground](https://github.com/letscontrolit/ESPEasyPluginPlayground) before we added it to ESPEasy), so until now there was no reason to change that 🤔

Gonzol999 commented:
I don't care about such currents, but this connector allows you to select a current in the range [1-1022 A], but there's still no option to change the shunt, and that's where I see a lack of consistency. A while ago, a similar build for the ESP8266 and INA226 was posted on the forum, with the option to change the shunt to an external one. Again, I don't care about that.

Gonzol999 commented:
I was referring to shunts like these, of course, for larger currents. Because looking for a resistor for, say, 50A is a bad idea.


https://www.aliexpress.us/item/3256804424770086.html?spm=a2g0o.productlist.main.1.632873732aYBoc&aem_p4p_detail=202512291139502324835655535800004893540&algo_pvid=0783ecf9-0244-4408-ac5c-10685ca9d8b6&algo_exp_id=0783ecf9-0244-4408-ac5c-10685ca9d8b6-0&pdp_ext_f=%7B%22order%22%3A%22198%22%2C%22eval%22%3A%221%22%2C%22fromPage%22%3A%22search%22%7D&pdp_npi=6%40dis%21PLN%2114.19%2114.19%21%21%2127.10%2127.10%21%402103890117670371904623939e7a09%2112000029826005658%21sea%21PL%21911166357%21X%211%210%21n_tag%3A-29919%3Bd%3Ae76cfd63%3Bm03_new_user%3A-29895&curPageLogUid=7VGqfUTJj1SP&utparam-url=scene%3Asearch%7Cquery_from%3A%7Cx_object_id%3A1005004611084838%7C_p_origin_prod%3A&search_p4p_id=202512291139502324835655535800004893540_1&gatewayAdapt=glo2usa4itemAdapt

tonhuisman commented:
Those shunts are for 75mV. Some of the INA chips, like INA219 and INA228, can handle higher shunt voltages, up to 320mV for INA219, and 160mV (fixed) for INA228. But that would still work fine, would just be a resolution-related issue.

tonhuisman commented:
> allows you to select a current in the range [1-1022 A], but there's still no option to change the shunt,

Well, I can limit that max. value to, say, 30A, I just added the max current setting as it is required for the new library I'm using.

Gonzol999 commented:
These currents don't bother me at all, so they can stay as they are.

tonhuisman committed:
[P132] Refactor Shunt settings and add more high-current options

tonhuisman committed:
[P132] Update documentation for extended Shunt setting

tonhuisman commented:
I've changed the resolution for the Shunt setting, and added more options for high-current shunts, to better match those to the Max. current setting.
@Gonzol999 Here's a Custom Energy build for ESP8266 for testing:
[ESP_Easy_mega_20251230_energy_ESP8266_4M1M.zip](https://github.com/user-attachments/files/24387902/ESP_Easy_mega_20251230_energy_ESP8266_4M1M.zip)

Gonzol999 commented:
I don't have a shunt at home because I don't measure such high currents. I downloaded your binary; in my opinion, it offers much more options. I'll get myself a shunt, let's say the smallest one, and run some tests. Thanks again.

tonhuisman committed:
[P132] Fix rounding for 400A shunt

tonhuisman committed:
[P132] Better selector-descriptions for shunt values

tonhuisman commented:
I've made a few minor adjustments, and built a fresh 2026 build: 😉 
[ESP_Easy_mega_20260101_energy_ESP8266_4M1M.zip](https://github.com/user-attachments/files/24402919/ESP_Easy_mega_20260101_energy_ESP8266_4M1M.zip)

Gonzol999 commented:
I have a 30A 75mV shunt and an INA226. Either I'm configuring something incorrectly or there's something wrong with the binary. Voltage is OK, but current measurement is a gamble in my opinion.

TD-er commented:
How is it connected?
Not sure how strict that chip is with its voltage levels, so maybe check if there is a voltage measurable between GND of the ESP and either pole of the shunt and especially if there is some voltage negative compared to GND of the ESP.

tonhuisman commented:
If the INA226 is installed on a board, any on-board/other shunt resistors have to be disconnected (or removed), and only the external shunt should be connected.

TD-er commented:
> If the INA226 is installed on a board, any on-board/other shunt resistors have to be disconnected (or removed), and only the external shunt should be connected.

Still, even if there is an existing shunt (in **parallel**), then the reported current should show a correlation to the actual current. Only lower.

Gonzol999 commented:
It's connected correctly. The original shunt resistor from the board has been removed. I loaded a binary I found on the forum that supports the INA226, and it works fine.

TD-er commented:
Do you have a link to that binary?

Gonzol999 commented:
https://github.com/letscontrolit/ESPEasy/issues/2622 
in this thread there is a link from which I downloaded the binary

This is an old link from November 18, 2021. Your request for the source code is also in this thread.

tonhuisman commented:
I have used the Zanduino INA library for adding support for the extra INAxxx models, though I haven't compared notes with the wollewald (INA226 explicit) library, as it is quite different.

What did you use for `Averaging samples` setting? As leaving that at 1 might turn out to be rather unstable. Also the `Conversion rate` settings may benefit the stability when allowing some more time for the conversion.

Gonzol999 commented:
Averaging samples: I tested probably all of them from this list, the closest current readings gave a value of 16. Conversion rate Current: I also tested all values ​​from the selection list, here the closest current consumption readings give a value of 1. Despite all this, the current measurement is unstable and significantly different from the actual one.

TD-er commented:
Where do you have the shunt? At the GND level or at the + side?
What voltage do you use?
Do you live in a 60Hz net frequency area or 50 Hz?
How long are the wires between the INA226 and the shunt?

In the [datasheet](https://www.ti.com/lit/ds/symlink/ina226.pdf) on page 14, there is some suggestion of filtering.
<img width="1309" height="643" alt="image" src="https://github.com/user-attachments/assets/12b3668e-68d6-438f-be0c-aa7ffb8c23a3" />

I'm discussing the code with Ton (via Signal) right now and we already found a few things that can be improved in the code.
So I guess Ton is already either writing this down for tomorrow or busy hacking the code :)

TD-er commented:
Also what current are you trying to measure?
Is it somewhat in the middle of the range of the shunt?
Or are you using a 30A shunt and try to measure 10 mA? (or some extreme like that)

I noticed a few places in the code of the library where quite a bit of resolution is lost due to careless calculations or completely illogical stripping of bits.

Gonzol999 commented:
<img width="1079" height="472" alt="ina226-schematic" src="https://github.com/user-attachments/assets/073f42ef-e544-4c63-b3d5-fc68e851a997" />

I've connected the shunt according to the attached diagram. The wires from the shunt to the INA are a few cm long (less than 10 cm). The voltage I'm trying to measure is 28 V, and these are external batteries connected to the UPS. The current from the mains should be 0 A under normal operation, but it can exceed 20 A under load. I live in Poland on a 240V 50 Hz grid.

The measurement results are very unstable, but I can probably check them with a filtering system at night. I'm just wondering why the binary I mentioned yesterday is practically OK.

tonhuisman commented:
I can't reach the home-garden.org site from my location (time-out), not sure if it still is available, or blocked for some reason, so there is no source-code available for comparison.
As TD-er said, there is room for improvement, so I'm working on that.

Gonzol999 commented:
Ok. I'll check with the filtering system. Another question: Has anyone else tested the INA226 with your input?

tonhuisman commented:
I tested with the on-board shunt and a quite low current. Haven't payed much attention to stable output, other than it stayed constant, as was to be expected, and changed when I changed the 'load'.

TD-er commented:
I'm personally not a fan of measuring current on the 'high side', but it should work as the voltage doesn't exceed the max. input of the INA226.
As a simple form of filtering, you could try to twist the wires from the shunt to the INA226 chip.
A UPS may add some magnetic interference to the signal and with twisted wires you can prevent that.

There are lots of reasons why the other binary file yielded more stable results.
For example Ton's code does not (yet) wait for a read cycle to be finished. Just assumes it is.
Also the load from the UPS on the batteries can be fluctuating a lot within 20 msec (1 cycle of 50 Hz), meaning the sensor should perform a number of measurements over a multiple of 10 msec (or 20) or maybe even a continuous reading.
Not sure what is possible here as I've not read the entire datasheet.
I know there are several 'integration time' settings and I've also seen some averaging parameters.
Maybe it best to configure the sensor to take some 1 msec readings and then average over 500 or 1000 samples. This would then result in more 'stable' readings, but on the other hand it will also result in some delay as every filtering introduces a delay in the response to changes.
Also it may require some tweaking as the shunt and voltage are read sequentially by the same ADC. So longer integration times can result in completely useless power calculations if the current fluctates a lot during a 50Hz cycle.

Apart from this, there are still some parts of the code that can be optimized as quite a lot of resolution can be lost how it is done right now.

le-lik commented:
Hello friends, you can add "ESP_Easy_mega_20251228_energy_ESP32_4M316k"RTC DS3231 and TFT LCD _P095_ILI9341 to the firmware, or tell me where to download the plugin with support for INA3231, INA226, etc., to try to build the firmware yourself.

tonhuisman commented:
We have documentation on how to set up a local development environment [here](https://espeasy.readthedocs.io/en/latest/Participate/PlatformIO.html). It is mostly describing how to develop a new plugin or feature, but also briefly explains how to build a Custom ESPEasy, containing exactly the plugins and features you want to use.

A much simpler solution, btw, is to get an ESP32 board with 8MB or 16MB of flash (ESP32 Classic, ESP32-S3, ESP32-C3 or ESP32-C6), and install one of the MAX builds that includes (nearly) all features and plugins.

le-lik commented:
Hi, thanks for the link.  And where can I download the max firmware?  I didn't come across any, or I was looking in the wrong place.

le-lik commented:
"tonhuisman" thank you for guiding me in the right direction, found the MAX firmware on your MAX forum, soldered the flash drive on my ESP32 Wroom to a 16M flash drive, filled in the firmware and everything worked.

tonhuisman commented:
> found the MAX firmware on your MAX forum

Not sure what you're talking about here, but official Release builds are available from our [Github Releases page](https://github.com/letscontrolit/ESPEasy/releases) (for flashing via a local tool) or from the WebFlasher at https://TD-er.nl/ESPEasy,
and the current development branch (`mega`) is irregularly made available at https://TD-er.nl/ESPEasy/latest

tonhuisman committed:
[P132] Improve calculation accuracy determining shunt microVolt factor

tonhuisman committed:
[P132] Only read measurement when conversion is ready

tonhuisman committed:
[P132] Enable extended features for Energy build ESP8266 for now

tonhuisman commented:
@Gonzol999 Here's a fresh build for you to test, it has the calculation improvement, mentioned above, and also checks if a measurement result is available before retrieving it, for improved stability of results.
[ESP_Easy_mega_20260125_energy_ESP8266_4M1M.zip](https://github.com/user-attachments/files/24846717/ESP_Easy_mega_20260125_energy_ESP8266_4M1M.zip)

NB: This build is based on the 20260108 release, _without_ the hotfix that was released today, as it's a test build. It will be updated to the `mega` development branch later.

Gonzol999 commented:
The current readings are stable but unrealistic. I'm testing a current of about 3 A, and the ESP with the current load shows 2732.02 A.
What is the purpose of the Max. Current: checkbox if there's a shunt resistor:?

TD-er commented:
Hmm sounds like a mixup of mA and uA somewhere. (or mV vs uV)

Gonzol999 commented:
Have you tested this yourself, or has anyone else done it? Maybe the problem is how I do it.

tonhuisman commented:
> The current readings are stable but unrealistic. I'm testing a current of about 3 A, and the ESP with the current load shows 2732.02 A. 

It's not obvious from either the configuration or the documentation, but this plugin shows the Current values in `mA`, as it always has. It seems that the value is in the expected range. You can use the Formula field to convert to `A`, if desired.

> What is the purpose of the Max. Current: checkbox if there's a shunt resistor:?

The actual current to measure doesn't have to match the capacity of the shunt, then only a part of the 75mV output is being used.
The Max. expected current setting (as [documented](https://github.com/letscontrolit/ESPEasy/pull/5462/files#diff-df36d17ef6fa0672fe4fb874bdd7a60c8e479e2f6c94dc21d7438abd3f8aff24R85-R86) but not yet published) is used to set the resolution as optimal as possible. If high currents are expected, then the range is larger, so resolution is lowered, but also reducing precision.

Gonzol999 commented:
Thanks for the replies.

If the plugin is supposed to return measurement results in mA, then it's probably OK. Yesterday, I solved the problem with the formula, dividing the measured value by 1000. You've clarified my doubts about the measurement range selection field. I'll continue testing your results with other settings.

TD-er commented:
Do you have other chips which are also supported by this plugin?
If so, I would greatly appreciate it if you could also try those on the same setup, just to make sure switching chip-models is also working properly and there isn't something else which is not intuitive.

Ton did move quite a lot of code around, so it is possible something got overlooked.

Gonzol999 commented:
I already have an INA228 at home, but I'm still waiting for the ESP32 from AliExpress. Testing would need to be done on the ESP32. For testing, I need to build a receiver with a more stable current source, preferably a resistor, because as you mentioned earlier, a UPS isn't the best receiver for such tests. But I can still test.

Gonzol999 commented:
Hello everyone.

I started testing the INA228 plug-in today, still with a low voltage of around 14V and a very low current of around 0.2A. The voltage measurements are OK, but the current is significantly off. The value measured by the plug-in fluctuates between 1.86 and 5.37mA. If, as Ton wrote earlier, the plug-in returns a value in mA, this result is outrageous. The board has a standard shunt resistor of 0.002Ω, and I have the Max. Current set to the lowest range, which is 1A. I noticed another problem: in the Tools/Interfaces tab, there's an I2C Scan tool. The Supported Devices field should indicate which module was found on the bus, but it's blank. The same is true for the INA226, so it's probably the plug-in's fault. One more question: INA228 modules have GPIO alerts. Do you plan to support this feature in the plug-in?

tonhuisman commented:
> but the current is significantly off.

Hm, that's somewhat unexpected. I recently (today) received an INA228. I'll hook that up and see what results I can achieve.

> ... I2C Scan tool. The Supported Devices field should indicate which module was found on the bus, but it's blank. The same is true for the INA226, so it's probably the plug-in's fault.

Depending on the build, many known device names are included. For ESP8266 builds these names are _not_ included for space reasons.

> One more question: INA228 modules have GPIO alerts. Do you plan to support this feature in the plug-in?

That's currently not planned.

tonhuisman committed:
[P132] Fix MQTT Discovery for Power setting

Gonzol999 commented:
> > ale prąd jest znacznie odchylony.
> 
> Hm, to trochę nieoczekiwane. Niedawno (dzisiaj) dostałem INA228. Podłączę go i zobaczę, jakie efekty uda mi się osiągnąć.

With our voltages, the discrepancies and instability in the reading are similar. It's possible that my INA228 board is of poor quality. How are your tests?
> 
> > ... Narzędzie skanujące I2C. Pole „Obsługiwane urządzenia” powinno wskazywać, który moduł został znaleziony na magistrali, ale jest puste. To samo dotyczy INA226, więc prawdopodobnie wina leży po stronie wtyczki.
> 
> W zależności od wersji, uwzględniono wiele znanych nazw urządzeń. W przypadku wersji ESP8266 nazwy te _nie_ są uwzględniane ze względu na brak miejsca.

This is not a problem, the found module address is returned and if someone connects the module to the I2C bus, they probably know what they are connecting
> 
> > Jeszcze jedno pytanie: moduły INA228 mają alerty GPIO. Czy planujecie obsługę tej funkcji we wtyczce?
> 
> Na razie nie ma takich planów.

this is also not a problem, similar functionality can be achieved using rules and e.g. the ESP GPIO port

sracing commented:
Hi, I am also desperately looking for INA226 integration, found this discussion here. Just ordered a ESP32-S3 with 16MB. Wanted to use the Max Firmware, since I have a couple of other sensors that need to be connected as well. However, I guess if I downlad the latest Release build from the Github Releases page, probably the P132 update for the INA 226 will not (yet) be in?

tonhuisman commented:
This PR isn't merged yet, so the code isn't included in any of the available builds.
You can try to create your own ESP32-S3 `MAX` build using this PR, though if you don't have any coding experience that can be a bit challenging (here's [documentation](https://espeasy.readthedocs.io/en/latest/Participate/PlatformIO.html) on how to set up your local development environment for ESPEasy). I won't be able to create a build for you today because of other obligations.

sracing commented:
You are right, no coding experience yet. So far I was always quite happy with the official builds. Guess I will have to wait.
In case you feel you need someone to test, let me know.

TD-er commented:
> You are right, no coding experience yet. So far I was always quite happy with the official builds. Guess I will have to wait. In case you feel you need someone to test, let me know.

I started a GH Actions build for this: https://github.com/letscontrolit/ESPEasy/actions/runs/21550809663
Should be ready in +/- 2 hours

tonhuisman commented:
@sracing 
The above build seems to have failed for some unknown reason, GH does that sometimes 🫣 
Here's the ESP32-S3 MAX build for testing (it's the build without PSRAM support, not sure if your board has that, may go into a boot-loop, then just report here and I'll upload the PSRAM build, ESP32-S3 is quite picky about that):
[ESP_Easy_mega_20260212_max_ESP32s3_16M8M.zip](https://github.com/user-attachments/files/25275026/ESP_Easy_mega_20260212_max_ESP32s3_16M8M.zip)
Has both the .factory.bin for uploading by using a flash tool at address 0 (initial  install), and the .bin for uploading via OTA.

sracing commented:
Wasted quite some time to find out why the flashing does not work till I noticed I got the non-PSRAM version this morning  (although I ordered the PSRAM version). So actually I was happy, cause I could use then the .bin from tonhuisman for uploading via OTA. I now have the Git Build "feature/P132-add-support-for-other-INA-chips_348c745" installed, but the P132 does not show a selector for the INA version (I have the INA226):

<img width="703" height="657" alt="image" src="https://github.com/user-attachments/assets/2909da3d-6aac-4f4a-b60b-9ea83a64903c" />

The device settings are actually not visible...

tonhuisman commented:
What if you enable the device (and Submit the page)?

NB: ESPEasy does not (in fact: shouldn't) communicate with the device until the plugin is enabled, and for auto-detection, it has to communicate with the device... 🤷‍♂️

sracing commented:
Same view. No Device Settings.
But I am actually checking the device if there is a issue. I´ve got a ESP32-S3-N16R8. I should have PSRAM, but I only manage to get the non-PASM firmware running. The PASM firmware does not boot. Give me a bit time to do some checks.

sracing commented:
OK, for reasons whatever, I now managed the PASM version to boot after installation. With the Git Build "feature/P132-add-support-for-other-INA-chips_348c745". Plugin is enabled. I get the same view, no device setting. Plugin is enabled.

The issue was that the I2C Adress was not yet configured in the Hardware tab (as this was a brand new board). Now Device Settings are visible.

Thanks for all the support, I will start testing phase over the weekend and provide feedback.

TD-er commented:
FYI
The ESP32-S3 with 16M flash and 8M PSRAM needs to access the PSRAM via an OPI SPI link.
The ESP32-S3 with less PSRAM need to access PSRAM with a QIO SPI link.
Sadly this cannot be changed at runtime, and if you try to access PSRAM via the wrong setup SPI link the device will bootloop.
That's why it is a safe bet to use a non-PSRAM build on the S3.

N.B. If you try to use addressable LEDs, like NeoPixel LEDs, you should not run an OPI SPI PSRAM build even if you have a matching board, as this PSRAM OPI SPI mode does interfere with how we try to address those LEDs and you cannot use those LEDs.  On those boards you need to run a non-PSRAM build.

See: https://espeasy.readthedocs.io/en/latest/ESPEasy/ESPchips.html#quad-octal-spi-mode

sracing commented:
Thats helpfull information, thanks. I am running the Git Build "feature/P132-add-support-for-other-INA-chips_348c745" from tonhuisman (non PSRAM built) on my ESP32-S3-N16R8 PSRAM since a few days now. The P132 with a INA226 (Standard 0.1 Ohm shunt version, max 0,81A measurement range) works flawless so far. My focus in the application is on the current output for a range of 50mA to 230mA. I made some tests with various sampling and conversion rates, but the standard 1 sample avg. and 1.1 msec already provides stable output under lab conditions. Will continue to do some more tests under field conditions next week, but the additional support for additional INA versions (at least for the INA 226 which I use) is definitively an enrichment.

sracing commented:
Just to add on my last post. P132 with a INA226 still runs stable. No issues during configuration or operations. The Sensor is placed in a area with short, but very strong EMI. A first sensor immediately showed a non-linear behaviour. I suspect the calibration register being impacted by EMI. A second sensor placed in a larger distance has no issues. But thats not related to the Plugin. As said, happy to see the extended plugin in the master.

TD-er commented:
Which part of your measurements is showing non-linear behavior?
Current measurement or voltage? (or both)

You may want to check if the shunt is getting warm and also if there may be other currents present.
Typically if there is some electrical connection between the power supply of the ESP + sensor and the current you try to measure, you have to make sure there is no potential difference between both grounds.

I think it is unlikely the I2C communication is disturbed, otherwise it would show up as highly 'unstable' readings (if any)

If the shunt is getting warm and the heat is not equally distributed, you may create some offset in the readings of microVolts (or even more) due to the thermocouple effect.
This is caused by joining different metals which create a voltage difference which is highly correlated by the temperature.
Usually this cancels out as this voltage is positive going from metal A to B and negative from B to A. However if there is somehow 1 side which is much warmer (e.g. bad contact), it can make a difference and for a shunt (which already has a low voltage drop) this low voltage offset may be enough to make the readings noticable non-linear.

sracing commented:
Appreciate the input. Actually voltage is not connected. Its the current measurement that shows stable, but non-linear outcome since it was placed close to a ignition wiring - before it was showing inconspicuous results. ESP & Sensor and the current to be measured are on different grounds. Not sure its getting warm, but as mentioned a new sensor now placed in bigger distance works fine since 1 week.

TD-er commented:
It could be that the EMI disturbances wipe some settings/data in the sensor, but then I would expect it to mess with the calibration points which are just a 2-point calibration. The ADC itself is still linear.

Or is this ignition wire 'firing' quite often, like from a "benzine" engine?
Then it is just messing with the readings.

sracing commented:
The inition firing is not constant during operations, just when starting the device - probably 15 impulses over 3 seconds. I am suspecting the calibration settings in the sensor. I tried to re-install and re-configure the plugin, but the linearity only came back instantly with a new sensor. So it seems to me the first sensor got somehow impacted.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P132-add-support-for-other-INA-chips
